src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,738 @@
+                {
+                    tree.ChildrenNodes.Add(nextDim, new List<Point>());
+                }
+                tree.ChildrenNodes[nextDim].Add(point);
+            }
+
+            if (aggNum == 0)
+            {
+                tree.Leaves.Add(point);
+            }
+        }
+
+        private BestDimension FindBestDimension(SortedDictionary<BestDimension, double> valueMap, Dictionary<BestDimension, double> valueRatioMap, double meanGain, bool isLeavesLevel = true)
+        {
+            BestDimension best = null;
+            foreach (KeyValuePair<BestDimension, double> dimension in valueMap)
+            {
+                if (dimension.Key.AnomalyDis.Count == 1 || (isLeavesLevel ? dimension.Value >= meanGain : dimension.Value <= meanGain))
+                {
+                    if (dimension.Key.AnomalyDis.Count > 1)
+                    {
+                        if (best == null || (!Double.IsNaN(valueRatioMap[best]) && (best.AnomalyDis.Count != 1 && (isLeavesLevel ? valueRatioMap[best].CompareTo(dimension.Value) <= 0 : valueRatioMap[best].CompareTo(dimension.Value) >= 0))))
+                        {
+                            best = dimension.Key;
+                        }
+                    }
+                    else
+                    {
+                        if (best == null || best.AnomalyDis.Count > 1)
+                        {
+                            best = dimension.Key;
+                        }
+                        else
+                        {
+                            if (!Double.IsNaN(valueRatioMap[best]) && (isLeavesLevel ? valueRatioMap[best].CompareTo(dimension.Value) <= 0 : valueRatioMap[best].CompareTo(dimension.Value) >= 0))
+                            {
+                                best = dimension.Key;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return best;
+        }
+
+        private double GetSurpriseScore(Point rootCausePoint, Point anomalyPoint)
+        {
+            double p;
+            double q;
+
+            if (anomalyPoint.ExpectedValue == 0)
+            {
+                p = 0;
+            }
+            else
+            {
+                p = rootCausePoint.ExpectedValue / anomalyPoint.ExpectedValue;
+            }
+
+            if (anomalyPoint.Value == 0)
+            {
+                q = 0;
+            }
+            else
+            {
+                q = rootCausePoint.Value / anomalyPoint.Value;
+            }
+
+            double surprise = 0;
+
+            if (p == 0)
+            {
+                surprise = 0.5 * (q * Log2(2 * q / (p + q)));
+            }
+            else if (q == 0)
+            {
+                surprise = 0.5 * (p * Log2(2 * p / (p + q)));
+            }
+            else
+            {
+                surprise = 0.5 * (p * Log2(2 * p / (p + q)) + q * Log2(2 * q / (p + q)));
```

Comment:
Add a comment to describe what the idea is here?

Reply:
> Add a comment to describe what the idea is here?

Have updated

Reply:
I don't see a comment. Can you please point out where or update it?

---
In reply to: [421337714](https://github.com/dotnet/machinelearning/pull/4925#discussion_r421337714) [](ancestors = 421337714)

Reply:
> I don't see a comment. Can you please point out where or update it?
> 
> In reply to: [421337714](https://github.com/dotnet/machinelearning/pull/4925#discussion_r421337714) [](ancestors = 421337714)

I have added the detailed idea of surprise score in mark down file and imported the file in the comments, you can see it on the top of this function.